compiler/hash-typecheck/src/exhaustiveness/lower.rs

```diff
@@ -135,123 +54,319 @@ impl<'tc> LowerPatOps<'tc> {
+                            .map(|p| self.construct_pat(*p))
+                            .collect_vec();
+
+                        // let args = self.builder().create_pat_args(children);
+                        // Pat::Tuple(args)
+
+                        todo!()
+                    }
+                    Term::Level1(Level1Term::NominalDef(_)) => {
+                        let _children = pat.fields.iter_patterns().map(|p| self.construct_pat(*p));
 
-                match nominal_def {
-                    NominalDef::Struct(_) => PatKind::Leaf { pats },
-                    // @@EnumToUnion: get the variant index here
-                    NominalDef::Enum(_) => unreachable!(),
+                        // let args = self.builder().create_pat_args(children);
+                        // Pat::Constructor(ConstructorPat { subject: pat.ty, args })
+                        todo!()
+                    }
+                    _ => tc_panic!(
+                        pat.ty,
+                        self,
+                        "Unexpected ty `{}` when converting to pattern",
+                        self.for_fmt(pat.ty),
+                    ),
+                }
+            }
+            DeconstructedCtor::IntRange(range) => self.construct_pat_from_range(pat.ty, range),
+            DeconstructedCtor::Str(_) => Pat::Lit(pat.ty),
+            DeconstructedCtor::List(List { kind }) => {
+                let children = pat.fields.iter_patterns().map(|p| self.construct_pat(*p));
+
+                match kind {
+                    ListKind::Fixed(_) => {
+                        let _inner_term = self.oracle().term_as_list(pat.ty).unwrap();
+
+                        // let inner = self.builder().create_pat_args(children);
+                        // Pat::List(ListPat { term: inner_term, inner })
+                        todo!()
+                    }
+                    #[allow(clippy::needless_collect)]
+                    ListKind::Var(prefix, _) => {
+                        let mut children = children.into_iter().peekable();
+
+                        // build the prefix and suffix components
+                        let prefix: Vec<_> = children.by_ref().take(prefix).collect();
+                        let suffix: Vec<_> = children.collect();
+
+                        // Create the `spread` dummy pattern
+                        let dummy = Pat::Spread(SpreadPat { name: None });
+                        let spread = self.pat_store().create(dummy);
+
+                        // Now create an inner collection of patterns with the inserted
+                        // spread pattern
+                        let _inner = prefix.into_iter().chain(once(spread)).chain(suffix);
+                        let _term = self.oracle().term_as_list(pat.ty).unwrap();
+
+                        // let elements = self.builder().create_pat_args(inner);
+                        // Pat::List(ListPat { term, inner: elements })
+                        todo!()
```

Comment:
todo left

Reply:
The todo is dependant on the `builder` having immutable access